### PR TITLE
Stack safe free-vars computation.

### PR DESCRIPTION
**Pull request by nickchapman-da**
Make the `freeVars` computation in `ClosureConversion` be stack-safe.
And test it is so!

Addresses an issue recorded in #11830

Part of the overarching issue: #11561

### CODE REVIEW

File: daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/ClosureConversion.scala [hunk 1]
```diff
@@ -319,50 +319,56 @@ private[speedy] object ClosureConversion {
...
-          bounds.zipWithIndex.foldLeft(go(body, bound + bounds.length, free)) {
-            case (acc, (expr, idx)) => go(expr, bound + idx, acc)
+  private[this] def freeVars(expr0: source.SExpr, depth0: Int): Set[Int] = {
+    @tailrec // woo hoo, stack safe!
+    def go(acc: Set[Int], work: List[(source.SExpr, Int)]): Set[Int] = {
+      // 'acc' is the (accumulated) set of free variables we have found so far.
+      // 'work' is a list of source expressions (paired with their depth) which we still have to process.
+      work match {
+        case Nil => acc // final result
+        case (expr, depth) :: work => {
```
Comment (remyhaemmerle-da) [inline daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/ClosureConversion.scala:332]:
```suggestion
        case (expr, depth) :: work =>
```

File: daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/ClosureConversion.scala [hunk 2]
```diff
@@ -319,50 +319,56 @@ private[speedy] object ClosureConversion {
...
+            case source.SEScopeExercise(body) =>
+              go(acc, (body, depth) :: work)
           }
-        case source.SELabelClosure(_, expr) =>
-          go(expr, bound, free)
-        case source.SETryCatch(body, handler) =>
-          go(body, bound, go(handler, 1 + bound, free))
-        case source.SEScopeExercise(body) =>
-          go(body, bound, free)
+        }
```
Comment (remyhaemmerle-da) [inline daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/ClosureConversion.scala:368]:
```suggestion
```

**Review by remyhaemmerle-da — APPROVED** [review]
Very Nice. Thanls

**Review by cocreature — APPROVED** [review]
Thanks!

File: daml-lf/interpreter/src/test/scala/com/digitalasset/daml/lf/speedy/ClosureConversionTest.scala [hunk 3]
```diff
@@ -57,59 +58,74 @@ class ClosureConversionTest extends AnyFreeSpec with Matchers with TableDrivenPr
...
-        ("Abs", abs1),
+        ("TryCatch2", tryCatch2),
+        ("scopeExercise", scopeExercise),
+        ("Labelclosure", labelClosure),
       )
     }
 
     {
-      val depth = 100000
+      val depth = 10000
```
[the remark this answers is not in the document]
Comment (cocreature) [reply daml-lf/interpreter/src/test/scala/com/digitalasset/daml/lf/speedy/ClosureConversionTest.scala:83]: seems sensible, thanks

**Pull request merged**